WiP: Support comment directives

janosh opened the pull request:
Continuing our discussion in #27.

janosh commented:
@andrewbranch Feel free to make changes yourself here as you see fit.

andrewbranch commented:
I’ve converted the TS to JS, moving the type definitions to a declaration file and referencing them in JSDoc. Also fixed a couple inconsistencies that TypeScript caught after getting the types set. What are you thinking for next steps?

andrewbranch commented on src/transformers/highlight-directive.js:
```diff
@@ -0,0 +1,53 @@
+ */
+const addClassName = (attrs, className) => {
+  // should we add a check here if attrs.className exists?
+  attrs.className += ` ${className}`;
```
`attrs.className` might not be defined yet. Let’s also return a new `attrs` object instead of mutating this one, just in case another transformer is holding a reference to it for some reason. (Mutating is probably slightly more efficient but it doesn’t really matter since this is happening at build time.)

janosh replied:
Done.

janosh commented:
Wrote [getCommentForLanguage](https://github.com/andrewbranch/gatsby-remark-vscode/pull/49/commits/3f671766acc4fe65681c22d72e0a605a3e3bc2e2#diff-05412402b415aaef71dc4db200d9dc13R18) with initial support for 17 common languages. Feel free to add more.

Also, wasn't sure how you'd like to JSDoc curried function.

andrewbranch commented:
This is looking great; thank you so much! Don’t worry too much about the JSDoc, I’ll fix up type annotations on anything that needs it later.

Is there anything else to do besides test it?

janosh commented:
> Is there anything else to do besides test it?

Nope, I'd say it's ready for testing. Perhaps that will reveal the need for some modifications here and there but other than that...

janosh commented:
I'll try and write some tests tomorrow. Is it possible to build with yarn? I tried `yarn && yarn build`. The latter did nothing:

```sh
yarn build
yarn run v1.19.1
$ node scripts/scrapeBuiltinExtensions.js
✨  Done in 0.14s.
```

Maybe add a `contribute.md`?

andrewbranch commented:
It didn’t do nothing, it just builds really fast 😄 

`yarn test` / `npm run test` is what you’re looking for. Good idea on adding a contributing guide 👍

janosh commented:
Refactored `getCommentForLanguage`, added a bunch more common languages and implemented
> [allow an option to be passed in to merge with that map](https://github.com/andrewbranch/gatsby-remark-vscode/issues/27#issuecomment-546633695)

I'm having trouble with the tests. Even when checking out upstream master and running `yarn test` after building, I get a lot of fails.
```sh
 › 1 snapshot failed.
 › 1 snapshot obsolete.
   • sets highlighted line class names 1
Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or run `ya
rn test -u` to update them.
 › 1 snapshot obsolete from 1 test suite. To remove it, run `yarn test -u`.
   ↳ test/gatsby-remark-vscode.test.js
       • sets highlighted line class names 1

Test Suites: 1 failed, 1 passed, 2 total
Tests:       19 failed, 13 passed, 32 total
Snapshots:   1 failed, 1 obsolete, 1 total
Time:        3.299s
Ran all test suites.
```
Could you add the tests?

andrewbranch commented:
Oh, woops, we forgot about the part that actually renders the attrs returned by the transformer onto HTML tags 🤦‍♂

andrewbranch commented:
Ok, that got more complicated than I intended it to, but we now use the attributes returned by transformers. I wrote some integration tests:

    ```ts
    const x = 0;
    const y = 1; // highlight-line
    const z = 2;
    ```

    ```shell
    npm install
    npm run build # highlight-line
    npm start
    ```

    ```js
    // This is where the interesting part is
    // highlight-start
    function somethingSuperInteresting() {
      console.log(this);
    }
    // highlight-end

    somethingSuperInteresting();
    ```

    ```swift
    enum Thing {
      case one
      // highlight-next-line
      case two(_ name: String)
    }
    ```

and it came out like this:

<img width="523" alt="image" src="https://user-images.githubusercontent.com/3277153/67831965-571f7680-fa9d-11e9-9df1-b9b7b4599245.png">

So it looks like the `highlight-next-line` and `highlight-start`/`highlight-end` scenarios aren’t quite working yet, but I’m calling it a night at the moment. We’re almost there!

janosh commented:
> Ok, that got more complicated than I intended it to, but we now use the attributes returned by transformers.

Wow, that looks like a lot of work! Thank you very much!

I pushed a possible fix for the stateful directives. I think we weren't passing state correctly from one line to the next. Unfortunately, I can't test it on my end so this is just coding in the dark and may not help. 🕶

andrewbranch commented:
Woohoo!

<img width="536" alt="image" src="https://user-images.githubusercontent.com/3277153/67916691-c5267500-fb53-11e9-9c1a-24896151cd95.png">

There are a few pieces of housekeeping, of varying levels of relevance to this actual change, that I need to take care of before publishing to npm:

- [ ] I want to reimplement the code-fence-options-based highlighting code as a line transformer as a way to further vet this API. It’ll also clean up index.js a bit, which is nice.
- [ ] I want to export the default set of line transformers as an array so that people can spread it if they want to add their own.
- [ ] We need docs for the comment-based line highlighting. I’m fine to ship without general line transformer docs.
- [ ] I need to update my GitHub Actions to the new version, because that’s how this thing gets published.

I’ll probably aim to get all this done this weekend. Feel free to hack on any of it in the meantime if you want, but no pressure. Also, but unrelated, I’d like to try to get the test infrastructure more solid—it’s just jest, so I’m surprised it didn’t work for you. I might try to get more details from you on your environment later if you have time.

Thanks again! 💯

andrewbranch commented:
I’m gonna go ahead and merge this and work on GitHub Actions in a new branch. Thanks so much for your work on this! One note FYI: I decided to remove the `languageCommentMap` option from the main plugin options. I started to document it in the README and it was a bit confusing, and it seems like a real edge case that someone will want to customize it. I left it as an option passed to `createHighlightDirectiveLineTransformer`, though, so in the rare case that someone needs it, they could construct the line transformer themselves. I also changed the array plugin option `lineTransformers` to `getLineTransformers`, which is passed (almost) all the other plugin options, so it will be easier for us and end-users to construct other options-dependent line transformers in the future.

Thanks again! 🎉

janosh commented:
I should be the one thanking you! This was fun and I'm really looking forward to using highlight directives now. 😀 

> I might try to get more details from you on your environment later if you have time.

Sure, hit me up any time.

andrewbranch commented:
I realized what happened with the tests when I hit it setting up CI too. All the default themes and languages come from the vscode submodule, which isn’t cloned by default. Maybe I should add it to a pre-build script.

janosh commented:
Awesome. Just [ported my personal site](https://github.com/janosh/janosh.io/commit/b8d140167b4cbb251e471e0bdb13b2ed80ae3ef1) to comment directives. Works like a charm! 😄
